src/combinatorics/rado/auxiliary.lean

```diff
@@ -0,0 +1,144 @@
+/-
+Copyright (c) 2023 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+
+import data.finset.card
+import data.fintype.basic
+
+/-!
+# Some auxiliary lemmas on finsets etc.
+
+These should probably go into `data.finset.basic`, `data.finset.image` or `data.fintype.basic`.
+-/
+
+namespace finset
+
+/-!
+Define versions of `eq_on`, `inj_on` and `restrict` for `finset`s.
+
+This is mostly for convenience so that we can use dot notation for these on `finset`s.
+-/
+
+section finset_versions
+
+variables {α : Type*} {β : Type*}
+
+/-- The restriction of `f` to a `finset` `s` -/
+def restrict (s : finset α) (f : α →  β) (a : ↥s) : β := f a
+
+/-- The statement that two functions `f` and `g` agree on a `finset` `s` -/
+def eq_on (f g : α → β) (s : finset α) : Prop := (s : set α).eq_on f g
+
+/-- The statement that a function `f` is injective on a `finset` `s` -/
+def inj_on (f : α → β) (s : finset α) : Prop := (s : set α).inj_on f
+
+-- We don't need the next two, but include them for completeness
+
+/-- The statement that a function `f` is surjective from a `finset` `s` to a `finset` `t` -/
+def surj_on (f : α → β) (s : finset α) (t : finset β) : Prop := (s : set α).surj_on f t
+
+/-- The statement that a function `f` is bijective from a `finset` `s` to a `finset` `t` -/
+def bij_on (f : α → β) (s : finset α) (t : finset β) : Prop := (s : set α).bij_on f t
+
+end finset_versions
+
+/-!
+Auxiliary `finset` lemmas.
+-/
+
+section α
+
+variables {α : Type*} [decidable_eq α]
+
+lemma insert_sdiff_singleton_self {a : α} {s : finset α} (h : a ∈ s) : s = insert a (s \ {a}) :=
+begin
+  ext b,
+  simp only [mem_insert, mem_sdiff, mem_singleton],
+  by_cases H : b = a; simp [h, H],
+end
+
+lemma insert_sdiff_singleton_self' {a : α} {s : finset α} (h : a ∉ s) : s = insert a s \ {a} :=
+begin
+  rw insert_sdiff_of_mem _ (mem_singleton_self _),
+  exact (sdiff_singleton_not_mem_eq_self _ h).symm,
+end
+
+lemma card_insert_eq_card_iff (a : α) (s : finset α) : (insert a s).card = s.card ↔ a ∈ s :=
+begin
+  rw card_insert_eq_ite,
+  simp only [imp_false, ite_eq_left_iff, nat.succ_ne_self, not_not],
```

Comment:
```suggestion
  rw [card_insert_eq_ite, ite_eq_right (nat.succ_ne_self _)],
```

Reply:
`ite_eq_right` is also unknown. There is `ite_eq_right_iff` (and shouldn't it be `left` anyway?).

Reply:
`by simp only [card_insert_eq_ite, imp_false, ite_eq_left_iff, nat.succ_ne_self, not_not]` works.